libyul/optimiser/RedundantStoreEliminator.cpp

```diff
@@ -0,0 +1,452 @@
+		applyOperation(op);
+
+	ControlFlowSideEffects sideEffects;
+	if (auto builtin = m_dialect.builtin(_functionCall.functionName.name))
+		sideEffects = builtin->controlFlowSideEffects;
+	else
+		sideEffects = m_controlFlowSideEffects.at(_functionCall.functionName.name);
+
+	if (!sideEffects.canContinue)
+	{
+		changeUndecidedTo(State::Unused, Location::Memory);
+
+		changeUndecidedTo(
+			sideEffects.canTerminate ?
+			State::Used :
+			State::Unused,
+			Location::Storage
+		);
+	}
+}
+
+void RedundantStoreEliminator::operator()(FunctionDefinition const& _functionDefinition)
+{
+	ScopedSaveAndRestore storeOperations(m_storeOperations, {});
+	RedundantStoreBase::operator()(_functionDefinition);
+}
+
+
+void RedundantStoreEliminator::operator()(Leave const&)
+{
+	changeUndecidedTo(State::Used);
+}
+
+void RedundantStoreEliminator::visit(Statement const& _statement)
+{
+	using evmasm::Instruction;
+
+	RedundantStoreBase::visit(_statement);
+
+	auto const* exprStatement = get_if<ExpressionStatement>(&_statement);
+	if (!exprStatement)
+		return;
+
+	FunctionCall const* funCall = get_if<FunctionCall>(&exprStatement->expression);
+	if (!funCall)
+		return;
+	optional<Instruction> instruction = toEVMInstruction(m_dialect, funCall->functionName.name);
+	if (!instruction)
+		return;
+
+	if (!ranges::all_of(funCall->arguments, [](Expression const& _expr) -> bool {
+		return get_if<Identifier>(&_expr) || get_if<Literal>(&_expr);
+	}))
+		return;
+
+	if (
+		*instruction == Instruction::SSTORE ||
+		(!m_ignoreMemory && (
+			*instruction == Instruction::EXTCODECOPY ||
+			*instruction == Instruction::CODECOPY ||
+			*instruction == Instruction::CALLDATACOPY ||
+			*instruction == Instruction::RETURNDATACOPY ||
+			*instruction == Instruction::MSTORE ||
+			*instruction == Instruction::MSTORE8
+		))
+	)
+	{
+		m_stores[YulString{}].insert({&_statement, State::Undecided});
+		vector<Operation> operations = operationsFromFunctionCall(*funCall);
+		yulAssert(operations.size() == 1, "");
+		m_storeOperations[&_statement] = move(operations.front());
+	}
+}
+
+vector<RedundantStoreEliminator::Operation> RedundantStoreEliminator::operationsFromFunctionCall(
+	FunctionCall const& _functionCall
+) const
+{
+	using evmasm::Instruction;
+
+	YulString functionName = _functionCall.functionName.name;
+	SideEffects sideEffects;
+	if (BuiltinFunction const* f = m_dialect.builtin(functionName))
+		sideEffects = f->sideEffects;
+	else
+		sideEffects = m_functionSideEffects.at(functionName);
+
+	optional<Instruction> instruction = toEVMInstruction(m_dialect, functionName);
+	if (!instruction)
+	{
+		vector<Operation> result;
+		// Unknown read is worse than unknown write.
+		if (sideEffects.memory != SideEffects::Effect::None)
+			result.emplace_back(Operation{Location::Memory, Effect::Read, {}, {}});
+		if (sideEffects.storage != SideEffects::Effect::None)
+			result.emplace_back(Operation{Location::Storage, Effect::Read, {}, {}});
```

Comment:
There is no plan to improve this because `Operation` contains references to local variables that denote the start and length of the operation. I think it would be too difficult to get this error-free if you want it to work across functions.

Reply:
Even in SSA form? If variables are never reassigned they could just as well be treated as globals. I guess recursion might be a problem but we already deal with loops which seems like a similar problem.

Reply:
That is true, but if you don't make the connection through the function parameters, then you cannot really reason about whether or not stores conflict with each other, can you?